kafka-cron/consumer/main.go

```diff
@@ -0,0 +1,138 @@
+	"os/signal"
+	"sync"
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+
+	"encoding/json"
+
+	"github.com/Shopify/sarama"
+	"github.com/berkeli/kafka-cron/types"
+	"github.com/google/uuid"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	brokerList = kingpin.Flag("brokerList", "List of brokers to connect").Default("localhost:9092").Strings()
+	topic      = kingpin.Flag("topic", "Topic name").Default("jobs-cluster-a").String()
+	retryTopic = kingpin.Flag("retryTopic", "Retry topic name").Default("jobs-cluster-a-retries").String()
+	partition  = kingpin.Flag("partition", "Partition number").Default("0").Int32()
+)
+
+func main() {
+
+	InitMonitoring(2112)
+
+	kingpin.Parse()
+	config := sarama.NewConfig()
+	config.Consumer.Return.Errors = true
+	brokers := *brokerList
+	master, err := sarama.NewConsumer(brokers, config)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	config = sarama.NewConfig()
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(*brokerList, config)
+	if err != nil {
+		log.Panic(err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+
+	defer func() {
+		if err := master.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+	consumer, err := master.ConsumePartition(*topic, *partition, sarama.OffsetNewest)
+	if err != nil {
+		log.Panic(err)
+	}
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt)
+	chDone := make(chan bool)
+	wgWorkers := sync.WaitGroup{}
+	go func() {
+		for {
+			select {
+			case err := <-consumer.Errors():
+				log.Println(err)
+			case msg := <-consumer.Messages():
+				var cmd types.Command
+				err := json.Unmarshal(msg.Value, &cmd)
+				if err != nil {
+					log.Println(err)
+				}
+				wgWorkers.Add(1)
+				// TODO: Add a worker pool with semaphore? What if jobs are dependent on each other?
+				processCommand(producer, cmd, &wgWorkers)
+			case <-signals:
+				chDone <- true
+				return
+			}
+		}
+	}()
+	<-chDone
+	log.Println("Interrupt is detected, shutting down gracefully...")
+	wgWorkers.Wait()
+}
+
+func processCommand(producer sarama.SyncProducer, cmd types.Command, wg *sync.WaitGroup) {
+	defer wg.Done()
+	log.Println("Starting a job for: ", cmd.Description)
+	//metrics
+	timer := prometheus.NewTimer(JobDuration.WithLabelValues(*topic, cmd.Description))
+	defer timer.ObserveDuration()
+	JobStatus.WithLabelValues("new", *topic, cmd.Description).Inc()
+
+	out, err := executeCommand(cmd.Command)
+	if err != nil {
+
+		log.Printf("Command: %s resulted in error: %s\n", cmd.Command, err)
+		if cmd.MaxRetries > 0 {
+			JobStatus.WithLabelValues("retry", *topic, cmd.Description).Inc()
+			cmd.MaxRetries--
+			log.Printf("Retrying command: %s, %d retries left\n", cmd.Command, cmd.MaxRetries)
+			cmdBytes, err := json.Marshal(cmd)
+
+			if err != nil {
+				log.Println(err)
+			}
+			_, _, err = producer.SendMessage(&sarama.ProducerMessage{
```

Comment:
It feels like there are three approaches to retries that make sense:
1. Retry in-line on the consumer that did the initial dequeue
2. Re-submit to a single topic of all consumers
3. Re-submit to a retry-specific topic with retry-specific consumers

What do you think are the pros/cons of each?

Reply:
I suggested a specific retry topic in the project. https://github.com/CodeYourFuture/immersive-go-course/tree/main/kafka-cron#part-3-handling-errors
In order to avoid failing job retries negatively impacting first-time runs, basically.
Also suggested it should be possible to configure consumers either to read from primary queue or retry queue which seems to be the case here.

Reply:
From my research, creating a separate queue for retries seems to be the standard (which is what I went with as Laura pointed out).
Pros/Cons for each:

1. Retry in-line on the consumer that did the initial dequeue
* CON: This adds additional code for retries (potentially with exponential backoff). 
* CON: Additional metrics to track retries
* CON: If it's a long running job, it could potentially block the 'normal' queue which we might be expecting to be done within certain time.
* PRO: No need to run several consumers
2. Re-submit to a single topic of all consumers
* PRO: Simple code, no need for a lot of additional code
* PRO: No need for multiple consumers
* PRO: 'normal' queue is respected
* CON: Additional metadata on messages if we want to keep track of failed/retries jobs
* CON: Additional metrics that we need to take into account
3. Re-submit to a retry-specific topic with retry-specific consumers
* PRO: Simple code, no need for a lot of additional code
* PRO: 'normal' queue is respected
* PRO: Easy to keep track of failed/retries jobs without much of custom metrics
* CON: Have to run multiple consumers

Reply:
That sounds about right :) I'd be curious to go a bit deeper, though, into the complexity of needing to run several consumers - why is that complex?